crates/ruff_graph/src/collector.rs

```diff
@@ -0,0 +1,83 @@
+use ruff_python_ast::helpers::collect_import_from_member;
+use ruff_python_ast::name::QualifiedName;
+use ruff_python_ast::visitor::source_order::walk_module;
+use ruff_python_ast::visitor::source_order::{walk_expr, walk_stmt, SourceOrderVisitor};
+use ruff_python_ast::{self as ast, Expr, Mod, Stmt};
+use ruff_python_stdlib::identifiers::is_identifier;
+
+/// Collect all imports for a given Python file.
+#[derive(Default, Debug)]
+pub(crate) struct Collector<'ast> {
+    /// Whether to detect imports from string literals.
+    string_imports: bool,
+    /// The collected imports from the Python AST.
+    imports: Vec<CollectedImport<'ast>>,
+}
+
+impl<'ast> Collector<'ast> {
+    pub(crate) fn new(string_imports: bool) -> Self {
+        Self {
+            string_imports,
+            imports: Vec::new(),
+        }
+    }
+
+    #[must_use]
+    pub(crate) fn collect(mut self, module: &'ast Mod) -> Vec<CollectedImport<'ast>> {
+        walk_module(&mut self, module);
+        self.imports
+    }
+}
+
+impl<'ast> SourceOrderVisitor<'ast> for Collector<'ast> {
+    fn visit_stmt(&mut self, stmt: &'ast Stmt) {
+        match stmt {
+            Stmt::ImportFrom(ast::StmtImportFrom {
+                names,
+                module,
+                level,
+                range: _,
+            }) => {
+                let module = module.as_deref();
+                let level = *level;
+                for alias in names {
+                    let qualified_name = collect_import_from_member(level, module, &alias.name);
+                    self.imports
+                        .push(CollectedImport::ImportFrom(qualified_name));
+                }
+            }
+            Stmt::Import(ast::StmtImport { names, range: _ }) => {
+                for alias in names {
+                    let qualified_name = QualifiedName::user_defined(&alias.name);
+                    self.imports.push(CollectedImport::Import(qualified_name));
+                }
+            }
+            _ => {
+                walk_stmt(self, stmt);
+            }
+        }
+    }
+
+    fn visit_expr(&mut self, expr: &'ast Expr) {
+        if self.string_imports {
+            if let Expr::StringLiteral(ast::ExprStringLiteral { value, range: _ }) = expr {
+                // Determine whether the string literal "looks like" an import statement: contains
+                // a dot, and consists solely of valid Python identifiers.
+                let value = value.to_str();
+                if value.split('.').all(is_identifier) {
```

Comment:
An empty string is also not a valid module name. You could fix that by doing something like this?

```diff
diff --git a/crates/red_knot_python_semantic/src/module_name.rs b/crates/red_knot_python_semantic/src/module_name.rs
index 885c6adf9..3bce5b2ec 100644
--- a/crates/red_knot_python_semantic/src/module_name.rs
+++ b/crates/red_knot_python_semantic/src/module_name.rs
@@ -59,7 +59,7 @@ impl ModuleName {
     }
 
     #[must_use]
-    fn is_valid_name(name: &str) -> bool {
+    pub fn is_valid_name(name: &str) -> bool {
         !name.is_empty() && name.split('.').all(is_identifier)
     }
 
diff --git a/crates/ruff_graph/src/collector.rs b/crates/ruff_graph/src/collector.rs
index 9b8c6108a..687fd7c20 100644
--- a/crates/ruff_graph/src/collector.rs
+++ b/crates/ruff_graph/src/collector.rs
@@ -1,9 +1,9 @@
+use red_knot_python_semantic::ModuleName;
 use ruff_python_ast::helpers::collect_import_from_member;
 use ruff_python_ast::name::QualifiedName;
 use ruff_python_ast::visitor::source_order::walk_module;
 use ruff_python_ast::visitor::source_order::{walk_expr, walk_stmt, SourceOrderVisitor};
 use ruff_python_ast::{self as ast, Expr, Mod, Stmt};
-use ruff_python_stdlib::identifiers::is_identifier;
 
 /// Collect all imports for a given Python file.
 #[derive(Default, Debug)]
@@ -64,7 +64,7 @@ impl<'ast> SourceOrderVisitor<'ast> for Collector<'ast> {
                 // Determine whether the string literal "looks like" an import statement: contains
                 // a dot, and consists solely of valid Python identifiers.
                 let value = value.to_str();
-                if value.split('.').all(is_identifier) {
+                if ModuleName::is_valid_name(value) {
                     let qualified_name = QualifiedName::from_dotted_name(value);
                     self.imports.push(CollectedImport::Import(qualified_name));
```

Reply:
My preference here is to simply call `ModuleName::new` and return the module name instead of the `QualifiedName`. It's not a valid module name if `ModuleName::new` returns `None`, otherwise it's a legal python module name.

This also avoids exposing any internal `ModuleName` methods.